presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,39 +1594,197 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
+
+        int numPaths = polygon.getPathCount();
+        boolean includeEdgeStartPoint = false;
+        for (int i = 0; i < numPaths; i++) {
+            int pathStartIndex = polygon.getPathStart(i);
+            int pathEndIndex = polygon.getPathEnd(i);
+
+            // check the last edge first
+            Point edgeStart = polygon.getPoint(pathEndIndex - 1);
+            if (edgeStart.equals(point)) {
+                return false;
+            }
+
+            for (int j = pathStartIndex; j < pathEndIndex; j++) {
+                Point edgeEnd = polygon.getPoint(j);
+
+                // all vertices are excluded
+                if (edgeEnd.equals(point)) {
+                    return false;
+                }
+
+                // if this is a vertical edge, it can only add an even (and
+                // therefore uninformative) number of crossings, so don't bother checking
+                if (isVerticalEdge(edgeStart, edgeEnd)) {
+                    // if this is the last edge
+                    if (j == pathEndIndex - 1) {
+                        if (haveSameLongitude(edgeEnd.getX(), point.getX()) && liesInRange(edgeEnd.getY(), point.getY(), NORTH_POLE.getY())) {
+                            intersectionSum++;
+                            continue;
+                        }
+                    }
+
+                    includeEdgeStartPoint = true;
+                    edgeStart = edgeEnd;
+                    continue;
+                }
+
+                // if the previous edge and the current edge share a vertex (i.e. not separated by a vertical edge),
+                // skip checking it this time to avoid double counting
+                // (if we didn't do this, consider the case of an upside down
+                // triangle with an external point directly below its tip)
+                if (!includeEdgeStartPoint && haveSameLongitude(edgeStart.getX(), point.getX())) {
+                    edgeStart = edgeEnd;
+                    continue;
+                }
+
+                if ((includeEdgeStartPoint && haveSameLongitude(edgeStart.getX(), point.getX())) &&
+                        Double.compare(edgeStart.getY(), point.getY()) >= 0) {
+                    edgeStart = edgeEnd;
+                    intersectionSum++;
+                    includeEdgeStartPoint = false;
+                    continue;
+                }
+
+                // the garden variety case
+                int nextEdgeEndpointIndex = (j + 1) % pathEndIndex;
+                if (edgeIntersectsTestEdge(edgeStart, edgeEnd, point, NORTH_POLE) &&
+                        !longitudeGrazesVertex(point.getX(), edgeStart, edgeEnd, polygon.getPoint(nextEdgeEndpointIndex))) {
+                    intersectionSum++;
+                }
+                includeEdgeStartPoint = false;
+                edgeStart = edgeEnd;
+            }
+        }
+        return intersectionSum % 2 == 1;
+    }
+
+    private static class SphericalCalculator
```

Comment:
This class is large enough, maybe it makes sense to extract it and put it in `presto-geospatial-toolkit`?